### PR TITLE
Fix data races triggered by functional tests.

### PR DESCRIPTION
**Pull request by gabriel-bjg**
Thread sanitizer triggers data races on functional tests. Steps to reproduce:

```
./configure --enable-debug --prefix=`pwd`/depends/x86_64-pc-linux-gnu/ --with-sanitizers=thread
make clean
make -j8
python3 test/functional/test_runner.py -j 4
```

Fixes:

- Function CWallet::KeepKey requires locking as it has concurrent access to database and member nKeysLeftSinceAutoBackup.
- Avoid data race when reading setInventoryTxToSend size by locking the read. If locking happens after the read, the size may change.
- Lock cs_mnauth when reading verifiedProRegTxHash.
- Make fRPCRunning atomic as it can be read/written from different threads simultaneously.
- Make m_masternode_iqr_connection atomic as it can be read/written from different threads simultaneously.
- Use a recursive mutex to synchronize concurrent access to quorumVvec.
- Make m_masternode_connection atomic as it can be read/written from different threads simultaneously.
- Make m_masternode_probe_connection atomic as it can be read/written from different threads simultaneously.
- Use a recursive mutex for synchronized access to activeMasterNode.

### CODE REVIEW

File: src/llmq/quorums.cpp [hunk 1]
```diff
@@ -517,10 +528,12 @@ size_t CQuorumManager::GetQuorumRecoveryStartOffset(const CQuorumCPtr pQuorum, c
     });
     std::sort(vecProTxHashes.begin(), vecProTxHashes.end());
     size_t nIndex{0};
-    for (size_t i = 0; i < vecProTxHashes.size(); ++i) {
-        if (activeMasternodeInfo.proTxHash == vecProTxHashes[i]) {
-            nIndex = i;
-            break;
+    {
+        for (size_t i = 0; i < vecProTxHashes.size(); ++i) {
+            if (WITH_LOCK(activeMasternodeInfoCs, return activeMasternodeInfo.proTxHash == vecProTxHashes[i])) {
```
Comment (PastaPastaPasta) [inline src/llmq/quorums.cpp:533]: Should we just lock outside of the loop here instead? And if we don't, then should remove redundant scope outside of loop

File: src/llmq/quorums.cpp [hunk 2]
```diff
@@ -781,9 +794,13 @@ void CQuorumManager::StartQuorumDataRecoveryThread(const CQuorumCPtr pQuorum, co
 
         while (nDataMask > 0 && !quorumThreadInterrupt) {
 
-            if (nDataMask & llmq::CQuorumDataRequest::QUORUM_VERIFICATION_VECTOR && pQuorum->quorumVvec != nullptr) {
-                nDataMask &= ~llmq::CQuorumDataRequest::QUORUM_VERIFICATION_VECTOR;
-                printLog("Received quorumVvec");
+            {
+                LOCK(pQuorum->quorumVvecCs);
+                if (nDataMask & llmq::CQuorumDataRequest::QUORUM_VERIFICATION_VECTOR &&
+                    WITH_LOCK(pQuorum->quorumVvecCs, return pQuorum->quorumVvec != nullptr)) {
```
Comment (PastaPastaPasta) [inline src/llmq/quorums.cpp:800]: this is a duplicate lock no?

File: src/llmq/quorums.cpp [hunk 3]
```diff
@@ -211,13 +221,15 @@ void CQuorumManager::TriggerQuorumDataRecoveryThreads(const CBlockIndex* pIndex)
...
-            const bool fWeAreQuorumMember = pQuorum->IsValidMember(activeMasternodeInfo.proTxHash);
+            const bool fWeAreQuorumMember = WITH_LOCK(activeMasternodeInfoCs, return pQuorum->IsValidMember(activeMasternodeInfo.proTxHash));
             const bool fSyncForTypeEnabled = mapQuorumVvecSync.count(pQuorum->qc->llmqType) > 0;
             const QvvecSyncMode syncMode = fSyncForTypeEnabled ? mapQuorumVvecSync.at(pQuorum->qc->llmqType) : QvvecSyncMode::Invalid;
             const bool fSyncCurrent = syncMode == QvvecSyncMode::Always || (syncMode == QvvecSyncMode::OnlyIfTypeMember && fWeAreQuorumTypeMember);
 
-            if ((fWeAreQuorumMember || (fSyncForTypeEnabled && fSyncCurrent)) && pQuorum->quorumVvec == nullptr) {
-                nDataMask |= llmq::CQuorumDataRequest::QUORUM_VERIFICATION_VECTOR;
+            {
+                if ((fWeAreQuorumMember || (fSyncForTypeEnabled && fSyncCurrent)) && WITH_LOCK(pQuorum->quorumVvecCs, return pQuorum->quorumVvec == nullptr)) {
```
Comment (PastaPastaPasta) [inline src/llmq/quorums.cpp:230]: No need for redundant scope if we use WITH_LOCK right?

File: src/llmq/quorums.cpp (hunk 1 above)
Comment (gabriel-bjg) [reply src/llmq/quorums.cpp]: We should just lock outside, fixed it.

File: src/llmq/quorums.cpp (hunk 2 above)
Comment (gabriel-bjg) [reply src/llmq/quorums.cpp]: yes, fixed it.

File: src/llmq/quorums.cpp (hunk 3 above)
Comment (gabriel-bjg) [reply src/llmq/quorums.cpp]: Right.

**Review by PastaPastaPasta — APPROVED** [review]
utACK

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
Thanks for finding and fixing these issues! 👍 

Few notes:
- it's better to avoid holding locks for too long to lower the risk of introducing deadlocks at some point (d876781e303458bf4c056b03055e12d06ed8b3f2)
- we should always lock `quorumVvecCs` whenever we access `quorumVvec` now that it's `GUARDED_BY(quorumVvecCs)` (099909adc56a3d0228d73d51776eb8455f619686)
- running `p2p_quorum_data.py` revealed a similar issue on my machine, this time for `skShare`  (da442281ca9611afc53d99d30ba1025302a9ffb1)

**Review by PastaPastaPasta — DISMISSED** [review]
re-utACK

File: src/masternode/activemasternode.cpp [hunk 4]
```diff
@@ -201,10 +202,11 @@ bool CActiveMasternodeManager::GetLocalAddress(CService& addrRet)
     if (!fFoundLocal) {
         bool empty = true;
         // If we have some peers, let's try to find our local address from one of them
-        g_connman->ForEachNodeContinueIf(CConnman::AllNodes, [&fFoundLocal, &empty](CNode* pnode) {
+        auto service = WITH_LOCK(activeMasternodeInfoCs, return activeMasternodeInfo.service);
+        g_connman->ForEachNodeContinueIf(CConnman::AllNodes, [&](CNode* pnode) {
             empty = false;
             if (pnode->addr.IsIPv4())
-                fFoundLocal = GetLocal(activeMasternodeInfo.service, &pnode->addr) && IsValidNetAddr(activeMasternodeInfo.service);
+                fFoundLocal = GetLocal(activeMasternodeInfo.service, &pnode->addr) && IsValidNetAddr(service);
```
Comment (UdjinM6) [inline src/masternode/activemasternode.cpp:209]:
```suggestion
                fFoundLocal = GetLocal(service, &pnode->addr) && IsValidNetAddr(service);
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Comment by PastaPastaPasta** [comment]
Needs rebase

File: src/masternode/activemasternode.cpp (hunk 4 above)
Comment (gabriel-bjg) [reply src/masternode/activemasternode.cpp]: Applied this and also your previous changes:

- done d876781, excluding changes where we pass temporary copies by values and do possibly do "old "data reads with no benefit. I kept only the changes that could trigger dead locks on the network and moved locks out of for loops that are quite big.  
- done 099909a
- done da44228

File: src/llmq/quorums.cpp [hunk 5]
```diff
@@ -197,8 +209,10 @@ void CQuorumManager::TriggerQuorumDataRecoveryThreads(const CBlockIndex* pIndex)
 
         // First check if we are member of any quorum of this type
         bool fWeAreQuorumTypeMember{false};
-        for (const auto& pQuorum : vecQuorums) {
-            if (pQuorum->IsValidMember(activeMasternodeInfo.proTxHash)) {
+
+        auto proTxHash = WITH_LOCK(activeMasternodeInfoCs, return activeMasternodeInfo.proTxHash);
+        for (const auto &pQuorum : vecQuorums) {
```
Comment (UdjinM6) [inline src/llmq/quorums.cpp:214]:
```suggestion
        for (const auto& pQuorum : vecQuorums) {
```

>Align pointers and references to the left i.e. use `type& var` and not `type &var`.

https://github.com/dashpay/dash/blob/master/doc/developer-notes.md#coding-style-c

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Review by UdjinM6 — APPROVED** [review]
utACK

File: src/llmq/quorums.cpp (hunk 5 above)
Comment (gabriel-bjg) [reply src/llmq/quorums.cpp]: applied

**Comment by gabriel-bjg** [comment]
> more fixes coming?
yes, i finished all fixes now, waiting for the builds.

**Comment by PastaPastaPasta** [comment]
There was any additional data race / deadlock I detected on a run of feature_block.py which I forwarded to @gabriel-bjg 

My understanding is he's planning to fix that and push

**Comment by gabriel-bjg** [comment]
> There was any additional data race / deadlock I detected on a run of feature_block.py which I forwarded to @gabriel-bjg
> 
> My understanding is he's planning to fix that and push

Regarding the data races, they are all triggered by zmq and libzmq. According to bitcoin tsan exceptions, zmq and libzmq data races are all suppressed (https://github.com/bitcoin/bitcoin/blob/master/test/sanitizer_suppressions/tsan). As agreed with @PastaPastaPasta we will ignore them for now and won't fix anything related to them in this PR.

The possible deadlock triggered by tsan is also a false positive. I present below the scenario which triggers the lock order inversion:
```
Thread19:
    LOCK2(cs_main, mempool.cs) (miner.cpp:235) => scriptcheckqueue->m_control_mutex (validation.cpp:2160) => cs_main, cgovernancemanager.cs (governance/governance.cpp:352)

Thread18:
    LOCK(cs_main) (validation.cpp:984) => LOCK(mempool.cs) (txmempool.cpp:1232)
```
Basically what tsan detects is T19 locking: cs_main (first locking), **mempool.cs**, m_control_mutex, **cs_main** (second locking as it's already locked, but it's a recursive mutex), cgovernancemanager.cs while T18 is locking: cs_main, mempool.cs. The possible deadlock detection comes from the fact that tsan is not able to detect that cs_main has been already locked before locking mempool.cs on thread19. So it sees thread19 locking mempool.cs, m_control_mutex, cs_main while it sees thread18 locking cs_main and mempool.cs afterwards.

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: src/evo/mnauth.cpp [hunk 6]
```diff
@@ -223,15 +212,15 @@ void CMNAuth::NotifyMasternodeListChanged(bool undo, const CDeterministicMNList&
...
+                if ((it->second.fields & CDeterministicMNStateDiff::Field_pubKeyOperator) && it->second.state.pubKeyOperator.GetHash() != pnode->GetVerifiedPubKeyHash()) {
                     doRemove = true;
                 }
             }
         }
 
         if (doRemove) {
             LogPrint(BCLog::NET_NETCONN, "CMNAuth::NotifyMasternodeListChanged -- Disconnecting MN %s due to key changed/removed, peer=%d\n",
-                     pnode->verifiedProRegTxHash.ToString(), pnode->GetId());
+                     pnode->GetVerifiedProRegTxHash().ToString(), pnode->GetId());
```
[the remark this answers is not in the document]
Comment (gabriel-bjg) [reply src/evo/mnauth.cpp:223]: fixed

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for squash merge
